# Machine Learning major project-(predicting credict risk)

* Name : Aravalam Rajasekhar
* Course : B.tech(3rd year)
* College : Sri Venkateswara University Of College Of engingineering,Tirupati
* Phone no:9849692369


# Understanding the problem statement

 Based on the user's data given by bank,the model has to be built such that to predict the customer defaulter on EMI payment 

# Data Volume

* There are 17521 rows and 2395 features in Training Data
* 20442 rows and 2394 features in Testing Data

# importing Libraries

In [332]:
import pandas as pd,numpy as np,fancyimpute
from sklearn.impute import IterativeImputer
import seaborn as sns
import matplotlib.pyplot as plt

# Importing the dataset

In [333]:
data_set=pd.read_csv('/content/sample_data/Train.csv')

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:3326: DtypeWarning: Columns (746,835) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


In [334]:
data_set.head()

Col1  Col2          Col3          Col4          Col5          Col6  \
0    RIGHGSN     1  1.086711e+04  8.648345e+03  1.576618e+04  2.890466e+04   
1  RIMHI5ZGP     0  2.584312e+05  2.179633e+05  1.866287e+05  1.866287e+05   
2  RIHOG1SZU     0  1.459355e+07  1.260060e+07  1.100315e+07  1.010315e+07   
3  RIHOLMQWU     0  6.727043e+05  8.181116e+05  6.794893e+05  6.794893e+05   
4  RIHO584ET     0 -1.218742e+06 -3.005361e+06 -1.666241e+05 -2.403574e+06   

           Col7          Col8          Col9         Col10  ...  Col2388  \
0  5.273655e+03  2.346153e+04  1.285597e+04  2.871509e+04  ...        0   
1  2.154519e+05  1.923615e+05  1.726538e+05  1.666974e+05  ...        0   
2  1.879882e+07  1.514035e+07  5.320778e+06  6.906324e+06  ...        0   
3  8.257254e+05  7.605803e+05  3.577134e+05  4.236612e+05  ...        0   
4  2.583183e+06  2.086546e+06  5.938358e+06  5.446532e+06  ...        0   

   Col2389  Col2390  Col2391  Col2392  Col2393  Col2394  Col2395  Col2396  \
0        1        0        1    214.0    214.0     68.0     22.0     22.0   
1        0        0        0     41.0     41.0     38.0     12.0     12.0   
2        1        1        0    378.0    378.0    310.0     21.0     21.0   
3        0        0        0    702.0    702.0    644.0    951.0    951.0   
4        0        0        0   1305.0   1305.0   1211.0   2205.0   2205.0   

   Col2397  
0      7.0  
1     11.0  
2     16.0  
3    743.0  
4   1831.0  

[5 rows x 2395 columns]

There are 2395 columns in training dataset

In [420]:
test_data_set=pd.read_csv('/content/sample_data/Test.csv')

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:3326: DtypeWarning: Columns (700,731,740,752,761,789,811,820,829,841,850) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


In [421]:
test_data_set.head()

Col1          Col3          Col4          Col5          Col6  \
0  RIGD58ZWD  8.167140e+04  8.614542e+04  7.532296e+04  7.532296e+04   
1  RIH660YDS -1.362824e+07 -4.493043e+06 -3.777257e+06 -4.493043e+06   
2  RIH660Q96  1.302700e+06  1.341449e+06  1.240085e+06  1.341449e+06   
3  RIYDO15W1  1.932258e+05  1.454433e+05  1.929148e+05  1.454433e+05   
4  RIYBGC1ZD  3.913385e+05  1.987337e+05  1.703432e+05  1.987337e+05   

           Col7          Col8          Col9         Col10         Col11  ...  \
0  1.125174e+05  1.002778e+05  2.038915e+04  3.053853e+04  8.395911e+04  ...   
1 -4.749764e+06 -4.749764e+06  1.147556e+07  1.147556e+07  1.272055e+07  ...   
2  1.477503e+06  1.477503e+06  3.904086e+05  3.904086e+05  4.979106e+05  ...   
3  1.318250e+04  1.318250e+04  8.650108e+04  8.650108e+04  5.896693e+03  ...   
4  3.046536e+05  3.046536e+05  2.528307e+05  2.528307e+05  3.034449e+05  ...   

   Col2388  Col2389  Col2390  Col2391  Col2392  Col2393  Col2394  Col2395  \
0        0        0        0        0     44.0     44.0     23.0     41.0   
1        1        1        2        2   1234.0   1234.0    680.0    154.0   
2        0        0        0        0     33.0     24.0     19.0      3.0   
3        0        0        0        0      8.0      8.0      3.0      7.0   
4        0        0        0        0      7.0      7.0      7.0      6.0   

   Col2396  Col2397  
0     41.0     36.0  
1    153.0     86.0  
2      1.0      1.0  
3      7.0      3.0  
4      6.0      5.0  

[5 rows x 2394 columns]

There are 2394 columns in testing data

# About target variable and independent features in training dataset

* **Col2** is target variable and remaining 2394 are independent features






In [336]:
len(data_set['Col2388'].unique())

5

# Shape of the train data

In [337]:
data_set.shape

(17521, 2395)

# Shape of test data

In [338]:
test_data_set.shape

(20442, 2394)


* dataset consists of 17521 rows and 2395 columns

In [339]:
x=data_set.columns
x

Index(['Col1', 'Col2', 'Col3', 'Col4', 'Col5', 'Col6', 'Col7', 'Col8', 'Col9',
       'Col10',
       ...
       'Col2388', 'Col2389', 'Col2390', 'Col2391', 'Col2392', 'Col2393',
       'Col2394', 'Col2395', 'Col2396', 'Col2397'],
      dtype='object', length=2395)

# Observing the data types of each columns

In [340]:
dict={}
for i in data_set.columns:
  if data_set[i].dtype in dict:
    dict[data_set[i].dtype]+=1
  else:
    dict[data_set[i].dtype]=1
dict

{dtype('O'): 3, dtype('int64'): 1548, dtype('float64'): 844}

# There are 3 columns of object datatype,1548 columns of int and 844 columns of type float

# Observation of Col1

In [341]:
len(data_set['Col1'].unique())

17521

# Col1 consists of categorical type unique values for 17521 rows which strongly supports the removal of Col1 column

# Removal of Col1 column from both train and test dataset

In [342]:
data_set.drop(['Col1'],axis=1,inplace=True)

In [343]:
test_data_set.drop(['Col1'],axis=1,inplace=True)

# % of credict risk cases (Col2=1) in train dataset

In [344]:
len(data_set[data_set['Col2']==1])

1761

In [345]:
(len(data_set[data_set['Col2']==1])/len(data_set))*100

10.050796187432224

# There are only 1761 rows out of 17521 rows which are credict risk cases(10%)

# No of columns consisting of null values:759

In [346]:
null_containing_columns=[]
for i in data_set.columns[1:]:
  if data_set[i].isnull().sum()!=0:
    null_containing_columns.append(i)
print("No of null value columns:",len(null_containing_columns))

No of null value columns: 759


# Displaying the name of the column and no of rows having null values in each of the columns

In [347]:
null_columns=pd.DataFrame(index=[null_containing_columns])
null_value_observations=[]
for i in null_containing_columns:
  null_value_observations.append(len(data_set[i][data_set[i].isnull()==True]))
null_columns['null_value_observations']=null_value_observations
null_columns

null_value_observations
Col13                       1249
Col14                       1244
Col15                       1246
Col16                       1243
Col17                       1470
...                          ...
Col2393                     1174
Col2394                     1186
Col2395                     1174
Col2396                     1174
Col2397                     1186

[759 rows x 1 columns]

# Displaying the column names whose missing value ratio>0.7

In [348]:
missing_value_ratio_columns,for_test_data=[],[]
for i in null_containing_columns:
    if len(data_set[i][data_set[i].isnull()==True])/len(data_set)>0.7:
      missing_value_ratio_columns.append(i)
      for_test_data.append(i)
print('No of columns whose missing value ratio>0.7 : ',len(missing_value_ratio_columns))
print('Missing value ratio>0.7 column names : \n')
missing_value_ratio_columns

No of columns whose missing value ratio>0.7 :  193
Missing value ratio>0.7 column names : 



['Col19',
 'Col20',
 'Col37',
 'Col38',
 'Col39',
 'Col40',
 'Col41',
 'Col99',
 'Col113',
 'Col138',
 'Col139',
 'Col140',
 'Col141',
 'Col142',
 'Col143',
 'Col144',
 'Col145',
 'Col146',
 'Col147',
 'Col148',
 'Col149',
 'Col150',
 'Col151',
 'Col152',
 'Col153',
 'Col154',
 'Col155',
 'Col156',
 'Col157',
 'Col158',
 'Col159',
 'Col171',
 'Col185',
 'Col249',
 'Col252',
 'Col253',
 'Col254',
 'Col259',
 'Col263',
 'Col265',
 'Col266',
 'Col271',
 'Col274',
 'Col275',
 'Col276',
 'Col281',
 'Col285',
 'Col287',
 'Col288',
 'Col335',
 'Col336',
 'Col338',
 'Col339',
 'Col340',
 'Col342',
 'Col343',
 'Col344',
 'Col347',
 'Col348',
 'Col351',
 'Col352',
 'Col353',
 'Col356',
 'Col357',
 'Col359',
 'Col360',
 'Col361',
 'Col363',
 'Col364',
 'Col365',
 'Col368',
 'Col369',
 'Col372',
 'Col373',
 'Col374',
 'Col375',
 'Col382',
 'Col446',
 'Col449',
 'Col450',
 'Col451',
 'Col456',
 'Col460',
 'Col462',
 'Col463',
 'Col468',
 'Col471',
 'Col472',
 'Col473',
 'Col478',
 'Col482',
 'Col48

# There are 193 columns whose missing value ratio>0.7.Those has to be removed

# Removing the columns whose missing value ratio>0.7 from train and test dataset

In [349]:
while len(missing_value_ratio_columns)>0:
  for i in null_containing_columns:
    if len(data_set[i][data_set[i].isnull()==True])/len(data_set)>0.7:
      null_containing_columns.remove(i)
      missing_value_ratio_columns.remove(i)
      data_set.drop([i],axis=1,inplace=True)
      test_data_set.drop([i],axis=1,inplace=True)

In [350]:
data_set.shape

(17521, 2201)

# There are 2201 columns after removing the greater missing value ratio columns

In [351]:
list(data_set.columns.values[1:])

['Col3',
 'Col4',
 'Col5',
 'Col6',
 'Col7',
 'Col8',
 'Col9',
 'Col10',
 'Col11',
 'Col12',
 'Col13',
 'Col14',
 'Col15',
 'Col16',
 'Col17',
 'Col18',
 'Col21',
 'Col22',
 'Col23',
 'Col24',
 'Col25',
 'Col26',
 'Col27',
 'Col28',
 'Col29',
 'Col30',
 'Col31',
 'Col32',
 'Col33',
 'Col34',
 'Col35',
 'Col36',
 'Col42',
 'Col43',
 'Col44',
 'Col45',
 'Col46',
 'Col47',
 'Col48',
 'Col49',
 'Col50',
 'Col51',
 'Col52',
 'Col53',
 'Col54',
 'Col55',
 'Col56',
 'Col57',
 'Col58',
 'Col59',
 'Col60',
 'Col61',
 'Col62',
 'Col63',
 'Col64',
 'Col65',
 'Col66',
 'Col67',
 'Col68',
 'Col69',
 'Col70',
 'Col71',
 'Col72',
 'Col73',
 'Col74',
 'Col75',
 'Col76',
 'Col77',
 'Col78',
 'Col79',
 'Col80',
 'Col81',
 'Col82',
 'Col83',
 'Col84',
 'Col85',
 'Col86',
 'Col87',
 'Col88',
 'Col89',
 'Col90',
 'Col91',
 'Col92',
 'Col93',
 'Col94',
 'Col95',
 'Col96',
 'Col97',
 'Col98',
 'Col100',
 'Col101',
 'Col102',
 'Col103',
 'Col104',
 'Col105',
 'Col106',
 'Col107',
 'Col108',
 'Col109',
 'Col11

# Imputation of remaining missing value columns using iterative imputer for continuous data columns and mode of the values of categorical type columns

In [352]:
removing_categorical_data=data_set.drop(['Col2'],axis=1)
while len(null_containing_columns)>0:
  for i in null_containing_columns:
    if len(data_set[i].unique())<=10:
      data_set[i].fillna(data_set[i].mode()[0],inplace=True)
      null_containing_columns.remove(i)
    else:
      data_set[i][data_set[i].isnull()==True]=IterativeImputer().fit_transform(removing_categorical_data[i].values.reshape(-1,1))
      null_containing_columns.remove(i)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:8: SettingWithCopyWarning: 
A value is trying

In [353]:
print("No of null value columns:",len(null_containing_columns))

No of null value columns: 0


In [354]:
null_containing_columns=[]
for i in data_set.columns[1:]:
  if len(data_set[i][data_set[i].isnull()==True])!=0:
    null_containing_columns.append(i)
print("No of null value columns:",len(null_containing_columns))

No of null value columns: 0


# There are no null values in any of the columns after doing removal of some columns and imputation

# Categorical Independent Variables having only 0 and 1 values will be removed if any of its value frequency is >95% in it

In [355]:
categorical_variables=[]
for i in data_set.columns[1:]:
  if len(data_set[i].unique())<=2 and 0 in data_set[i].unique() and 1 in data_set[i].unique():
    categorical_variables.append(i)


# Number of categorical Variables

In [356]:
len(categorical_variables)

18

In [357]:
categorical_variables

['Col974',
 'Col1019',
 'Col1066',
 'Col1111',
 'Col1158',
 'Col1203',
 'Col1250',
 'Col1251',
 'Col1295',
 'Col1704',
 'Col1749',
 'Col1796',
 'Col1841',
 'Col1888',
 'Col1933',
 'Col1980',
 'Col1981',
 'Col2025']

# Frequency distribution of each of the categorical variable

In [358]:
for i in categorical_variables:
  print(data_set[i].value_counts())

0    17514
1        7
Name: Col974, dtype: int64
0    17517
1        4
Name: Col1019, dtype: int64
0    17514
1        7
Name: Col1066, dtype: int64
0    17517
1        4
Name: Col1111, dtype: int64
0    17514
1        7
Name: Col1158, dtype: int64
0    17517
1        4
Name: Col1203, dtype: int64
0    17514
1        7
Name: Col1250, dtype: int64
0    17517
1        4
Name: Col1251, dtype: int64
0    17517
1        4
Name: Col1295, dtype: int64
0    17514
1        7
Name: Col1704, dtype: int64
0    17517
1        4
Name: Col1749, dtype: int64
0    17514
1        7
Name: Col1796, dtype: int64
0    17517
1        4
Name: Col1841, dtype: int64
0    17514
1        7
Name: Col1888, dtype: int64
0    17517
1        4
Name: Col1933, dtype: int64
0    17514
1        7
Name: Col1980, dtype: int64
0    17517
1        4
Name: Col1981, dtype: int64
0    17517
1        4
Name: Col2025, dtype: int64


# Filtering categorical variables having frequency of any 1 value of >95%

In [359]:
high_frequency_variables=[]
for i in categorical_variables:
  for j in data_set[i].value_counts().index:
    if data_set[i].value_counts()[j]/len(data_set)>0.95:
      high_frequency_variables.append(i)
      break

    

In [360]:
high_frequency_variables

['Col974',
 'Col1019',
 'Col1066',
 'Col1111',
 'Col1158',
 'Col1203',
 'Col1250',
 'Col1251',
 'Col1295',
 'Col1704',
 'Col1749',
 'Col1796',
 'Col1841',
 'Col1888',
 'Col1933',
 'Col1980',
 'Col1981',
 'Col2025']

In [361]:
print('No of categorical variables having any value of frequency >95% : ',len(high_frequency_variables))

No of categorical variables having any value of frequency >95% :  18


# 18 categorical columns are having any 1 value with >95% frequency in it.So,It has to be removed because it is not properly influencing the target variable

In [362]:
data_set.drop(high_frequency_variables,axis=1,inplace=True) #dropping 18 categorical columns in both train & test dataset
test_data_set.drop(high_frequency_variables,axis=1,inplace=True)

In [363]:
data_set.shape

(17521, 2183)

# After removing categorical variables having value with >95% frequency,dataset columns are reduced to 2183 

# Removing rows having null values in test dataset

In [364]:
test_data_set.dropna(axis=0,inplace=True)

# Displaying the percentage of outliers in each column containing continuous values

In [365]:
dummy_dataframe=pd.DataFrame()
outlier_data_frame_columns,no_of_Outliers,p_of_Outliers=[],[],[]
for i in data_set.columns:
  if len(data_set[i].unique())>2:
    outlier_data_frame_columns.append(i)
    q1=data_set[i].quantile(0.25)
    q3=data_set[i].quantile(0.75)
    iqr=q3-q1
    lower_limit=q1-1.5*iqr
    upper_limit=q3+1.5*iqr
    def taking_outlier_values(data):
      if data<lower_limit or data>upper_limit:
        return data
      return 0
    dummy_dataframe['applied_values']=data_set[i].apply(taking_outlier_values)
    len_outlier_values=len(dummy_dataframe['applied_values'][dummy_dataframe['applied_values']!=0])
    p_of_outliers=int((len_outlier_values/len(data_set))*100)
    no_of_Outliers.append(len_outlier_values)
    p_of_Outliers.append(p_of_outliers)
outlier_data_frame=pd.DataFrame(index=[outlier_data_frame_columns])
outlier_data_frame['no_of_Outliers']=no_of_Outliers
outlier_data_frame['%_of_Outliers']=p_of_Outliers
outlier_data_frame







no_of_Outliers  %_of_Outliers
Col3               2169             12
Col4               2061             11
Col5               2072             11
Col6               2065             11
Col7               1955             11
...                 ...            ...
Col2393             915              5
Col2394             884              5
Col2395            1299              7
Col2396            1234              7
Col2397            1198              6

[2182 rows x 2 columns]

# No of columns containing outliers < 10%

In [366]:
print('No of columns containing outliers < 10% is ',len(outlier_data_frame[outlier_data_frame['%_of_Outliers']<10].index))

No of columns containing outliers < 10% is  1507


# There are 1507 columns which contains outliers < 10%

# Replacing less than 10% containing outlier values of the column with mean values

In [367]:
while len(outlier_data_frame[outlier_data_frame['%_of_Outliers']<10].index)>0:
  for i in outlier_data_frame[outlier_data_frame['%_of_Outliers']<10].index[0]:
    q1=data_set[i].quantile(0.25)
    q3=data_set[i].quantile(0.75)
    iqr=q3-q1
    lower_limit=q1-1.5*iqr
    upper_limit=q3+1.5*iqr
    def replacing_outliers(data):
      if data<lower_limit or data>upper_limit:
        return data_set[i].mean()
      return data
    data_set[i]=data_set[i].apply(replacing_outliers)
    outlier_data_frame.drop([i],axis=0,inplace=True)

/usr/local/lib/python3.7/dist-packages/pandas/core/generic.py:4150: PerformanceWarning: dropping on a non-lexsorted multi-index without a level parameter may impact performance.
  obj = obj._drop_axis(labels, axis, level=level, errors=errors)


# Analysing Correlations

In [368]:
corr_data_set=data_set.corr()

In [369]:
corr_data_set.sort_values('Col2',axis=1,ascending=False,inplace=True)
corr_data_set.sort_values('Col2',axis=0,ascending=False,inplace=True)

In [370]:
corr_data_set['Col2'][1:].max()

0.1491767001412036

In [371]:

def apply_color(val):
  font_color='black'
  background_color='white'
  if (val>=0.9 and val<1) or (val<=-0.9 and val>=-1):
    font_color='green'
    background_color='yellow'
  elif (val>=0.7 and val<0.9) or (val>-0.9 and val<=-0.7):
    font_color='green'
  elif (val>=0.4 and val<0.7) or (val>-0.7 and val<=-0.4):
    font_color='blue'
  elif val==1:
    font_color='black'
  else:
    font_color='red'
  return 'color: %s;background-color: %s' % (font_color,background_color)

# Correlation value between independent and target variable having >=0.1
# (values in green indicates Stronger Correlation,blue-moderate,red-Weak correlation)

In [372]:
corr_data_set_columns_01=list(corr_data_set[corr_data_set['Col2']>=0.1].index)
corr_data_set[corr_data_set_columns_01][corr_data_set['Col2']>=0.1].style.applymap(apply_color)

In [373]:
len(corr_data_set[corr_data_set['Col2']>=0.1])-1 #excluding target variable

60

# Observation in correlated values of Col2 >=0.1

* There are **60 independent variables** which are having correlation value >=0.1
* The highest correlation value of target value is **0.149177** with **Col431.**
* Top-5 independent variables having correlation values >=0.1 are **Col431,Col380,Col183,Col234,Col1502.**
* There are **21 multicollinearity features.**
* **Col380,Col183,Col1502,Col1322,Col577,Col1412,Col2232,Col2052,Col1592,Col2142,Col636,Col2322,Col1364,Col1544,Col1454,Col1634,Col627,Col2094,Col2274,Col2184,Col2364** are having very highly correlated values(>=0.9 or <=-0.9) leading to multicollinearity as we cannot decide which on feature the target variable is more dependent because of multicollinearity.
* **Col380,Col183,Col577** are multicollinear to each other
* **Col1502,Col1322,Col1412,Col2232,Col2052,Col1592,Col2142,Col2322,Col1364,Col1544,Col1454,Col1634,Col2094,Col2274,Col2184,Col2364** are havimg multicollinearity with each other.
* **Col636 and Col627** are multicollinear with each other

# Correlation value between independent and target variable having values between 0 and 0.1

In [374]:
corr_data_set_columns_001=list(corr_data_set[(corr_data_set['Col2']>=0) & (corr_data_set['Col2']<0.1)].index)
corr_data_set_columns_001.insert(0,'Col2')
corr_data_set[corr_data_set_columns_001][((corr_data_set['Col2']>=0) & (corr_data_set['Col2']<0.1))|(corr_data_set['Col2']==1)]

Col2    Col913   Col2296     Col75    Col615    Col166   Col2386  \
Col2     1.000000  0.099571  0.099432  0.099262  0.098793  0.098646  0.098559   
Col913   0.099571  1.000000  0.118543  0.097334  0.293182  0.193043  0.119598   
Col2296  0.099432  0.118543  1.000000  0.189896  0.044915  0.180196  0.995212   
Col75    0.099262  0.097334  0.189896  1.000000  0.048912  0.263706  0.188900   
Col615   0.098793  0.293182  0.044915  0.048912  1.000000  0.207328  0.043689   
...           ...       ...       ...       ...       ...       ...       ...   
Col1201  0.000078 -0.050465  0.098689  0.160071 -0.204353 -0.121525  0.098643   
Col1248  0.000078 -0.048823  0.099298  0.144130 -0.203846 -0.106574  0.100564   
Col2047  0.000054 -0.037416  0.039465  0.101886 -0.122373 -0.068272  0.038821   
Col168   0.000015 -0.027588  0.031546  0.011615 -0.060791 -0.001377  0.031779   
Col1017  0.000007 -0.050174  0.102814  0.157253 -0.206877 -0.120813  0.102866   

          Col1635    Col419    Col178  ...   Col1747    Col501   Col1230  \
Col2     0.098401  0.095613  0.095487  ...  0.000189  0.000157  0.000124   
Col913   0.137132  0.289774  0.203533  ... -0.050535  0.030125  0.026315   
Col2296  0.580459  0.033440  0.180176  ...  0.102684 -0.008750  0.075141   
Col75    0.182064  0.009787  0.249255  ...  0.155481 -0.073268  0.209641   
Col615   0.077827  0.908258  0.170671  ... -0.206847  0.047002  0.053870   
...           ...       ...       ...  ...       ...       ...       ...   
Col1201  0.113415 -0.258519 -0.099613  ...  0.976516 -0.111275  0.110281   
Col1248  0.115168 -0.236200 -0.088206  ...  0.719535 -0.083456  0.132610   
Col2047  0.040746 -0.145789 -0.052613  ...  0.087002 -0.095325  0.077096   
Col168   0.019331 -0.064289  0.006513  ...  0.058817 -0.044615 -0.011042   
Col1017  0.115878 -0.259950 -0.097146  ...  0.996958 -0.112960  0.110311   

           Col972    Col642   Col1201   Col1248   Col2047    Col168   Col1017  
Col2     0.000104  0.000090  0.000078  0.000078  0.000054  0.000015  0.000007  
Col913  -0.047113  0.006253 -0.050465 -0.048823 -0.037416 -0.027588 -0.050174  
Col2296  0.105837  0.005520  0.098689  0.099298  0.039465  0.031546  0.102814  
Col75    0.146334  0.032682  0.160071  0.144130  0.101886  0.011615  0.157253  
Col615  -0.210791  0.015157 -0.204353 -0.203846 -0.122373 -0.060791 -0.206877  
...           ...       ...       ...       ...       ...       ...       ...  
Col1201  0.726119 -0.027261  1.000000  0.714764  0.091770  0.050297  0.979997  
Col1248  0.955938 -0.026004  0.714764  1.000000  0.099578  0.062571  0.718774  
Col2047  0.098672  0.032087  0.091770  0.099578  1.000000  0.028346  0.089078  
Col168   0.072276 -0.010806  0.050297  0.062571  0.028346  1.000000  0.058603  
Col1017  0.744359 -0.028070  0.979997  0.718774  0.089078  0.058603  1.000000  

[928 rows x 928 columns]

# Observation in correlated value of target variable between 0 and 0.1
* These independent variables correlation with target variable are closer to zero.So,it is very negligibly influencing the target variable.
* There are 929 independent variables which are having the correlation values with target variable between 0 and 0.1
* The First independent variable having correlation value between 0 and 0.1 is Col913 having correlation value of 0.099571
* The top-5 independent features having values between 0 and 0.1 are Col913,Col2296,Col75,Col615,Col176
* Some independent variables are having multicollinearity(>=0.9 or <=-0.9)


# Correlation value between independent and target variable having < 0 and >=-0.1

In [375]:
corr_data_set_columns__010=list(corr_data_set[(corr_data_set['Col2'])<0 & (corr_data_set['Col2']>-0.1)].index)
corr_data_set_columns__010.insert(0,'Col2')
corr_data_set[corr_data_set_columns__010][((corr_data_set['Col2'])<0 & (corr_data_set['Col2']>-0.1))|(corr_data_set['Col2']==1)]

Col2   Col2314   Col1977   Col1442   Col1054   Col1558    Col169  \
Col2     1.000000 -0.000009 -0.000043 -0.000046 -0.000046 -0.000055 -0.000056   
Col2314 -0.000009  1.000000  0.105609  0.137438  0.137438  0.141657  0.007987   
Col1977 -0.000043  0.105609  1.000000  0.315702  0.315702  0.312350  0.006370   
Col1442 -0.000046  0.137438  0.315702  1.000000  1.000000  0.956071 -0.015317   
Col1054 -0.000046  0.137438  0.315702  1.000000  1.000000  0.956071 -0.015317   
...           ...       ...       ...       ...       ...       ...       ...   
Col931  -0.108514  0.215318  0.073385 -0.050010 -0.050010 -0.025140  0.031195   
Col932  -0.108901  0.216240  0.073398 -0.049758 -0.049758 -0.024854  0.031242   
Col16   -0.111779 -0.048802  0.087985 -0.036253 -0.036253 -0.029732  0.047004   
Col933  -0.112103  0.218861  0.067716 -0.054062 -0.054062 -0.029734  0.028370   
Col13   -0.129519 -0.053270  0.035932 -0.064993 -0.064993 -0.058311  0.042279   

          Col1839    Col421    Col493  ...    Col300    Col497    Col647  \
Col2    -0.000058 -0.000062 -0.000062  ... -0.101459 -0.102002 -0.103428   
Col2314  0.020341 -0.021301 -0.021301  ...  0.134157  0.134609  0.138690   
Col1977  0.182043 -0.015798 -0.015798  ...  0.097078  0.098685  0.088590   
Col1442  0.134366  0.260295  0.260295  ...  0.031459  0.032747  0.032107   
Col1054  0.134366  0.260295  0.260295  ...  0.031459  0.032747  0.032107   
...           ...       ...       ...  ...       ...       ...       ...   
Col931   0.018654 -0.201742 -0.201742  ...  0.544171  0.544717  0.522573   
Col932   0.018364 -0.201748 -0.201748  ...  0.543946  0.545623  0.523235   
Col16    0.319183 -0.103352 -0.103352  ...  0.071580  0.071848  0.050127   
Col933   0.018684 -0.186448 -0.186448  ...  0.523474  0.524430  0.554476   
Col13    0.248160 -0.080748 -0.080748  ...  0.065296  0.065240  0.052842   

           Col626    Col614    Col931    Col932     Col16    Col933     Col13  
Col2    -0.107731 -0.107941 -0.108514 -0.108901 -0.111779 -0.112103 -0.129519  
Col2314  0.048906 -0.046479  0.215318  0.216240 -0.048802  0.218861 -0.053270  
Col1977  0.065486  0.036099  0.073385  0.073398  0.087985  0.067716  0.035932  
Col1442  0.138296  0.306384 -0.050010 -0.049758 -0.036253 -0.054062 -0.064993  
Col1054  0.138296  0.306384 -0.050010 -0.049758 -0.036253 -0.054062 -0.064993  
...           ...       ...       ...       ...       ...       ...       ...  
Col931   0.327301  0.189997  1.000000  0.998381  0.144594  0.928986  0.131750  
Col932   0.327466  0.190274  0.998381  1.000000  0.144371  0.930509  0.131514  
Col16    0.089410  0.013756  0.144594  0.144371  1.000000  0.127946  0.914799  
Col933   0.346177  0.214293  0.928986  0.930509  0.127946  1.000000  0.117737  
Col13    0.097059  0.039218  0.131750  0.131514  0.914799  0.117737  1.000000  

[1196 rows x 1196 columns]

# Observation in correlated value of target variable between 0 and -0.1
* Like previous case,These independent variables are also having negligible correlation with target variable.
* 1192 independent variables are having the correlation values with target variable between 0 and -0.1
* The First independent variable having correlation value between 0 and -0.1 is Col2314 having correlation value of -0.000009
* The top-5 independent features having values between 0 and -0.1 are Col2314,Col1977,Col1054,Col1442,Col1558
* Some independent variables are having multicollinearity(>=0.9 or <=-0.9)

# Correlation value between independent and target variable having <= -0.1
# (values in green indicates Stronger Correlation,blue-moderate,red-Weak correlation)

In [376]:
corr_data_set_columns__01=list(corr_data_set[corr_data_set['Col2']<=-0.1].index)
corr_data_set_columns__01.insert(0,'Col2')
corr_data_set[corr_data_set_columns__01][(corr_data_set['Col2']<=-0.1) | (corr_data_set['Col2']==1)].style.applymap(apply_color)


In [377]:
len(corr_data_set[corr_data_set['Col2']<=-0.1])-1 #excluding target variable

10

# Observation in correlated values of Col2 <=-0.1
* These variables are impacting target variable negatively which if the value of these values increases,the value of of target variable decreases and vice-versa
* Only 10 independent variables are having correlation value <= -0.1
* The First Independent variable having correlation < -0.1 with target variable is Col430 with correlation value -0.100440
* Top-5 independent variables having correlation values < -0.1 with target variable are Col430,Col300,Col497,Col647,Col626.
* There are **8 multicollinearity features.**
* **Col300,Col497,Col647** are having multicollinearity with each other
* Similarly,**Col931,Col932,Col933** are having multicollinearity
* **Col13,Col16** are also having multicollinearity with each other


# Multicollinearity of independent variables(correlation>=0.9 or correlation<=-0.9)

In [378]:
multi_collinearity,multi_collinearity_values,corr_with_target={},{},{}
for c1 in data_set.columns[1:]:
  corr_with,corr_values,corr_value_target=[],[],0
  for c2 in data_set.columns[1:]:
    if c1!=c2 and (corr_data_set[c1][c2]>=0.9 or corr_data_set[c1][c2]<=-0.9):
      corr_with.append(c2)
      corr_values.append(round(corr_data_set[c1][c2],5))
      corr_value_target=round(corr_data_set['Col2'][c1],5)
      multi_collinearity[c1]=corr_with
      multi_collinearity_values[c1]=corr_values
      corr_with_target[c1]=corr_value_target

In [379]:
multi_collinearity_data_set=pd.DataFrame({'variable_name':multi_collinearity.keys(),'multicollinearity_with':multi_collinearity.values(),'respective_correlation_values':multi_collinearity_values.values()})
def count_multicollinear_features(val):
  return len(val)
multi_collinearity_data_set['no_of_multicollinearity_features']=multi_collinearity_data_set['multicollinearity_with'].apply(count_multicollinear_features)
multi_collinearity_data_set['correlation_with_target_variable']=corr_with_target.values()

In [380]:
multi_collinearity_data_set.sort_values('correlation_with_target_variable',ascending=False)

variable_name                             multicollinearity_with  \
219         Col380                                   [Col183, Col577]   
72          Col183                                   [Col380, Col577]   
1158       Col1502  [Col1322, Col1364, Col1412, Col1454, Col1544, ...   
978        Col1322  [Col1364, Col1412, Col1454, Col1502, Col1544, ...   
364         Col577                                   [Col183, Col380]   
...            ...                                                ...   
604         Col931                                   [Col932, Col933]   
605         Col932                                   [Col931, Col933]   
13           Col16                                            [Col13]   
606         Col933                                   [Col931, Col932]   
10           Col13                                            [Col16]   

                          respective_correlation_values  \
219                                  [0.99883, 0.91703]   
72                                   [0.99883, 0.91569]   
1158  [0.99803, 0.94601, 0.99887, 0.948, 0.94858, 0....   
978   [0.94633, 0.99915, 0.94813, 0.99803, 0.94758, ...   
364                                  [0.91569, 0.91703]   
...                                                 ...   
604                                  [0.99838, 0.92899]   
605                                  [0.99838, 0.93051]   
13                                             [0.9148]   
606                                  [0.92899, 0.93051]   
10                                             [0.9148]   

      no_of_multicollinearity_features  correlation_with_target_variable  
219                                  2                           0.14435  
72                                   2                           0.14416  
1158                                15                           0.13884  
978                                 15                           0.13876  
364                                  2                           0.13847  
...                                ...                               ...  
604                                  2                          -0.10851  
605                                  2                          -0.10890  
13                                   1                          -0.11178  
606                                  2                          -0.11210  
10                                   1                          -0.12952  

[2030 rows x 5 columns]

In [381]:
multi_collinearity_data_set.sort_values('no_of_multicollinearity_features',ascending=False)

variable_name                             multicollinearity_with  \
1342       Col1690  [Col940, Col960, Col984, Col1004, Col1032, Col...   
1430       Col1782  [Col940, Col960, Col1004, Col1032, Col1052, Co...   
1473       Col1826  [Col940, Col960, Col984, Col1004, Col1052, Col...   
721        Col1052  [Col940, Col960, Col984, Col1004, Col1032, Col...   
851        Col1188  [Col940, Col960, Col984, Col1004, Col1032, Col...   
...            ...                                                ...   
462         Col716                                           [Col805]   
461         Col715                                           [Col804]   
459         Col713                                           [Col802]   
1523       Col1879                                          [Col1149]   
2029       Col2396                                          [Col2395]   

                          respective_correlation_values  \
1342  [0.93891, 0.99187, 0.90138, 0.93114, 0.9347, 0...   
1430  [0.93465, 0.9822, 0.92517, 0.93247, 0.99139, 0...   
1473  [0.90069, 0.92418, 0.93817, 0.97882, 0.92208, ...   
721   [0.94432, 0.99174, 0.90245, 0.93089, 0.94118, ...   
851   [0.90863, 0.92962, 0.95254, 0.98533, 0.90296, ...   
...                                                 ...   
462                                           [0.91716]   
461                                           [0.92145]   
459                                           [0.93674]   
1523                                          [0.97624]   
2029                                          [0.98205]   

      no_of_multicollinearity_features  correlation_with_target_variable  
1342                                45                          -0.06572  
1430                                42                          -0.06590  
1473                                41                          -0.07273  
721                                 41                          -0.06340  
851                                 41                          -0.06946  
...                                ...                               ...  
462                                  1                          -0.00208  
461                                  1                          -0.06642  
459                                  1                          -0.00142  
1523                                 1                          -0.04342  
2029                                 1                           0.00324  

[2030 rows x 5 columns]

In [382]:
multicollinearity_variables=multi_collinearity_data_set['variable_name'][multi_collinearity_data_set['no_of_multicollinearity_features']>0]

In [383]:
len(multicollinearity_variables)

2030

# There are total 2030 variables which are multicollinear with one or more variables

#Feature Scaling(StandardScaler)
* It rescales the distribution of values so that the mean of the values is 0 and standard deviation =1

In [384]:
from sklearn.preprocessing import StandardScaler
for i in data_set.columns:
  if not(len(data_set[i].unique())==2 and 1 in data_set[i].unique() and 0 in data_set[i].unique()):
    data_set[i]=StandardScaler().fit_transform(data_set[i].values.reshape(-1,1))


# Dataset after performing StandardScaler

In [385]:
data_set

Col2       Col3       Col4       Col5       Col6       Col7       Col8  \
0         1  -0.027221  -0.016558  -0.015392   0.001659  -0.028708  -0.012773   
1         0   0.297979   0.276735   0.268525   0.260602   0.262729   0.212782   
2         0  19.128559  17.627378  18.242027  16.541017  26.030756  20.174791   
3         0   0.842167   1.117667   1.087498   1.069755   1.108945   0.971599   
4         0  -1.642434  -4.239809  -0.318465  -3.991857   3.545866   2.742334   
...     ...        ...        ...        ...        ...        ...        ...   
17516     0  -0.029142  -0.023455  -0.035399  -0.033570  -0.008052   0.009768   
17517     0   0.044726  -0.001900  -0.012315   0.012055  -0.029587  -0.025103   
17518     0   0.013950   0.015795   0.000600  -0.004110  -0.022084  -0.033367   
17519     0   0.051093   0.108847   0.121496   0.131527   0.146651   0.137407   
17520     0  -0.028108  -0.014462  -0.024734  -0.029140  -0.021642  -0.030256   

            Col9      Col10      Col11  ...   Col2388   Col2389   Col2390  \
0      -0.074791  -0.003339  -0.103582  ... -0.842135  0.642027 -0.877262   
1       0.659732   0.459468   0.632296  ... -0.842135 -0.777401 -0.877262   
2      24.323471  23.064872  35.723224  ... -0.842135  0.642027  0.461821   
3       1.510372   1.321351   1.509409  ... -0.842135 -0.777401 -0.877262   
4      27.162223  18.168579  13.157760  ... -0.842135 -0.777401 -0.877262   
...          ...        ...        ...  ...       ...       ...       ...   
17516  -0.110405  -0.083047  -0.060355  ... -0.842135 -0.777401 -0.877262   
17517   0.085604   0.101824  -0.098946  ...  0.514458 -0.777401  0.461821   
17518   0.066277   0.037771  -0.137862  ... -0.842135 -0.777401 -0.877262   
17519   0.129259   0.088370   0.144189  ... -0.842135  0.642027 -0.877262   
17520  -0.133483  -0.099360  -0.136970  ... -0.842135 -0.777401 -0.877262   

        Col2391   Col2392   Col2393   Col2394   Col2395   Col2396   Col2397  
0      0.582993  0.253705  0.253526  1.822091  0.982799  1.033878 -0.315311  
1     -0.815587 -0.208518 -0.196054  0.374730 -0.394129 -0.347904  0.549137  
2     -0.815587  0.253705  0.253526  0.238724  0.845106  0.895699  1.629695  
3     -0.815587  0.253705  0.253526  0.238724  0.431455  0.417600  0.410751  
4     -0.815587  0.253705  0.253526  0.238724  0.431455  0.417600  0.410751  
...         ...       ...       ...       ...       ...       ...       ...  
17516 -0.815587 -0.303886 -0.388519 -0.252460 -0.118743 -0.486082 -0.531422  
17517 -0.815587  0.253705  0.253526  2.883489  2.222034  0.895699  0.765248  
17518 -0.815587 -1.448313 -1.447079 -1.362104 -1.495671 -1.453330 -1.395869  
17519  0.582993 -0.399255 -0.420597 -0.927896 -0.669514 -0.762439 -0.963646  
17520 -0.815587 -1.448313 -1.447079 -1.362104 -1.495671 -1.453330 -1.179758  

[17521 rows x 2183 columns]

In [386]:
data_set5=data_set.copy()

In [387]:
from sklearn.linear_model import LogisticRegression

In [388]:
train_data_x5=data_set5[data_set5.columns[1:]]

In [389]:
train_data_y5=data_set5['Col2']

In [390]:
regr5=LogisticRegression().fit(train_data_x5,train_data_y5)

/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:818: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG,


In [391]:
from sklearn.metrics import confusion_matrix
confusion_matrix(train_data_y5,regr5.predict(train_data_x5))

array([[15615,   145],
       [ 1416,   345]])

# Dimensionality Reduction Technique

# creating new columns by adding features with high multicollinearity and removing those features

In [392]:
corr_data_set_columns_01=list(corr_data_set.index[1:])
old_dataset_len,perm_list,new_col=len(corr_data_set_columns_01),[],[]
for i in range(old_dataset_len):
  temp_multicollinear_list,bkn=[],0
  for j in range(i+1,old_dataset_len):
    if corr_data_set[corr_data_set_columns_01[i]][corr_data_set_columns_01[j]]>0.9 or corr_data_set[corr_data_set_columns_01[i]][corr_data_set_columns_01[j]]<-0.9:
      temp_multicollinear_list.append(corr_data_set_columns_01[j])
  if len(temp_multicollinear_list)>0:
    temp_multicollinear_list.insert(0,corr_data_set_columns_01[i])
    num=1
    stri=''
    for k in temp_multicollinear_list:
      stri+=k+'+'
    data_set[stri[:-1]]=0
    test_data_set[stri[:-1]]=0
    new_col.append(stri[:-1])
    for k in temp_multicollinear_list:
      data_set[stri[:-1]]+=data_set[k]
      test_data_set[stri[:-1]]+=test_data_set[k]
      perm_list.append(k)
data_set.drop(perm_list,axis=1,inplace=True)
test_data_set.drop(perm_list,axis=1,inplace=True)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:14: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead.  To get a de-fragmented frame, use `newframe = frame.copy()`
  
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead.  To get a de-fragmented frame, use `newframe = frame.copy()`
  from ipykernel import kernelapp as app


# New columns list

In [393]:
new_col

['Col380+Col183+Col577',
 'Col183+Col577',
 'Col1502+Col1322+Col1412+Col2232+Col2052+Col1592+Col2142+Col2322+Col1364+Col1544+Col1454+Col1634+Col2094+Col2274+Col2184+Col2364',
 'Col1322+Col1412+Col2232+Col2052+Col1592+Col2142+Col2322+Col1364+Col1544+Col1454+Col1634+Col2094+Col2274+Col2184+Col2364',
 'Col1412+Col2232+Col2052+Col1592+Col2142+Col2322+Col1364+Col1544+Col1454+Col1634+Col2094+Col2274+Col2184+Col2364',
 'Col2232+Col2052+Col1592+Col2142+Col2322+Col1364+Col1544+Col1454+Col1634+Col2094+Col2274+Col2184+Col2364',
 'Col2052+Col1592+Col2142+Col2322+Col1364+Col1544+Col1454+Col1634+Col2094+Col2274+Col2184+Col2364',
 'Col1592+Col2142+Col2322+Col1364+Col1544+Col1454+Col1634+Col2094+Col2274+Col2184+Col2364',
 'Col2142+Col2322+Col1364+Col1544+Col1454+Col1634+Col2094+Col2274+Col2184+Col2364',
 'Col636+Col627',
 'Col2322+Col1364+Col1544+Col1454+Col1634+Col2094+Col2274+Col2184+Col2364',
 'Col1364+Col1544+Col1454+Col1634+Col2094+Col2274+Col2184+Col2364',
 'Col1544+Col1454+Col1634+Col2094+Col22

# dataset after removing multicollinear columns and creating new columns

In [394]:
data_set

Col2     Col17     Col18     Col21     Col22     Col23     Col24  \
0         1 -0.300749 -1.143593  1.201416  1.217685  1.027555  1.057346   
1         0  0.899998  0.376949 -0.449729 -0.316640  0.665216  0.295517   
2         0  0.597656  0.060854 -1.329180 -1.159816 -0.092523 -0.237716   
3         0  1.380635 -1.054664 -1.293563 -1.115369 -0.309707 -0.700693   
4         0 -0.300749  0.919682  0.086271  0.476032  0.187280  1.671503   
...     ...       ...       ...       ...       ...       ...       ...   
17516     0 -1.334663 -0.266901  0.926227 -0.167440 -0.160087 -1.682626   
17517     0 -0.208095  0.919682  0.086271  2.178614  0.187280  0.518691   
17518     0  1.992745  0.919682  0.929224 -2.092431  0.187280 -0.777656   
17519     0 -1.334663 -1.089432 -0.946292 -1.039654 -0.687401 -0.593256   
17520     0 -1.334663 -0.731708 -2.312774 -2.049922 -2.532416 -1.845804   

          Col25     Col26     Col27  ...  Col1830+Col1738  \
0     -0.170446 -0.289401 -0.047720  ...        -1.976819   
1      1.765471 -0.157106  2.050510  ...         3.161797   
2      1.935743 -0.157106  2.050510  ...         3.161797   
3     -0.997034  1.046739 -3.093160  ...         3.161797   
4     -0.957063  2.166487  2.050510  ...         3.161797   
...         ...       ...       ...  ...              ...   
17516  0.097275  0.415905  0.031000  ...        -1.976819   
17517 -0.172813  0.021953 -0.003328  ...        -1.976819   
17518  0.302814  1.672868  2.050510  ...        -1.976819   
17519 -0.714251 -1.034157 -1.463888  ...         3.161797   
17520  0.908563 -1.123093  2.050510  ...        -1.976819   

       Col418+Col221+Col614  Col221+Col614  Col873+Col872  Col233+Col430  \
0                  6.052037       4.085573       1.236106       2.852842   
1                  0.159730       0.091316       1.236106       1.674926   
2                  3.773459       2.597536       4.662813       0.548960   
3                 -0.846029      -0.581356       1.236106       3.538766   
4                 -3.580949      -2.377506      -2.110341       1.041905   
...                     ...            ...            ...            ...   
17516             -4.016174      -2.635020      -1.187142      -2.660559   
17517             -0.235999      -0.155652      -3.563819      -1.852731   
17518             -5.433001      -3.577336       1.236106      -2.769445   
17519             -0.403018      -0.379097       0.651258       0.290789   
17520             -5.433001      -3.577336       1.236106      -2.769445   

       Col300+Col497+Col647  Col497+Col647  Col931+Col932+Col933  \
0                 -2.758260      -1.825493             -4.159392   
1                 -0.008729      -0.001011             -0.513007   
2                  1.280524       0.891033              4.683188   
3                  6.300950       4.142885              1.563741   
4                  2.214181       1.344302              1.711041   
...                     ...            ...                   ...   
17516             -2.758260      -1.825493             -4.159392   
17517             -2.758260      -1.825493             -3.215992   
17518              7.472450       4.940220              1.673614   
17519              3.089262       2.468349              5.419595   
17520             -2.758260      -1.825493             -4.159392   

       Col932+Col933  Col16+Col13  
0          -2.742098     0.685288  
1          -0.306270     4.150879  
2           3.172348    -0.034022  
3           0.993264    -0.034022  
4           1.024186    -0.034022  
...              ...          ...  
17516      -2.742098    -0.355767  
17517      -2.133825     2.054472  
17518       1.113663     1.524479  
17519       3.748345     2.881263  
17520      -2.742098    -0.085923  

[17521 rows x 1775 columns]

# FeatureScaling for new columns

In [395]:
for i in new_col:
  if not(len(data_set[i].unique())==2 and 1 in data_set[i].unique() and 0 in data_set[i].unique()):
    data_set[i]=StandardScaler().fit_transform(data_set[i].values.reshape(-1,1))

In [396]:
test_data_set.dropna(axis=0,inplace=True)

# Building the model

In [397]:
train_data_x=data_set[data_set.columns[1:]]
train_data_y=data_set['Col2']

# Using StackingClassifier to build the model
* This is an **ensemble learning method** which is used **increase the performance of the model**.
* Took DecisionTreeClassifier,RandomForestClassifier,KNeighborsClassifier,Supportvectormachinesclassifier into stackmodel

In [398]:
from sklearn.ensemble import StackingClassifier
from sklearn.tree import DecisionTreeClassifier
dtc=DecisionTreeClassifier()
from sklearn.ensemble import RandomForestClassifier
rfc=RandomForestClassifier()
from sklearn.neighbors import KNeighborsClassifier
knn=KNeighborsClassifier()
from sklearn.svm import SVC
svm=SVC()
lr=LogisticRegression()
clf=[('dtc',dtc),('rfc',rfc),('knn',knn),('svm',svm)] #Taking DecisionTreeClassifier,RandomForestClassifier,KNeighborsClassifier,Supportvectormachinesclassifier
stackmodel=StackingClassifier(estimators=clf,final_estimator=lr)
stackmodel.fit(train_data_x,train_data_y)

StackingClassifier(estimators=[('dtc', DecisionTreeClassifier()),
                               ('rfc', RandomForestClassifier()),
                               ('knn', KNeighborsClassifier()),
                               ('svm', SVC())],
                   final_estimator=LogisticRegression())

# Checking f1_score for StackingClassifier

In [399]:
confusionmatrix=confusion_matrix(train_data_y,stackmodel.predict(train_data_x))

In [400]:
precision=confusionmatrix[0][0]/(confusionmatrix[0][0]+confusionmatrix[0][1])
recall=confusionmatrix[0][0]/(confusionmatrix[0][0]+confusionmatrix[1][0])
f1_score_stacking=2*precision*recall/(precision+recall)
f1_score_stacking

0.995735270889275

# Accuracy for StackingClassifier

In [401]:
Accuracy_stacking=(confusionmatrix[0][0]+confusionmatrix[1][1])/(confusionmatrix[0][0]+confusionmatrix[1][1]+confusionmatrix[0][1]+confusionmatrix[1][0])
Accuracy_stacking

0.9922949603333143

In [402]:
stackmodel.predict(test_data_set[test_data_set.columns])

array([1, 1, 1, 1])

# predicting target variable for testing dataset using stackingclassifier

In [403]:
stackmodel_prediction=stackmodel.predict(test_data_set[test_data_set.columns])
stackmodel_prediction

array([1, 1, 1, 1])

# Using BaggingClassifier

* This method is used to **decrease the variance of the model**

In [404]:
from sklearn.ensemble import BaggingClassifier
from sklearn.tree import DecisionTreeClassifier
bag_model=BaggingClassifier(base_estimator=DecisionTreeClassifier(),n_estimators=10,max_samples=0.8,oob_score=True,random_state=0)
bag_model.fit(train_data_x,train_data_y)

/usr/local/lib/python3.7/dist-packages/sklearn/ensemble/_bagging.py:707: UserWarning: Some inputs do not have OOB scores. This probably means too few estimators were used to compute any reliable oob estimates.
  "Some inputs do not have OOB scores. "
/usr/local/lib/python3.7/dist-packages/sklearn/ensemble/_bagging.py:712: RuntimeWarning: invalid value encountered in true_divide
  oob_decision_function = predictions / predictions.sum(axis=1)[:, np.newaxis]


BaggingClassifier(base_estimator=DecisionTreeClassifier(), max_samples=0.8,
                  oob_score=True, random_state=0)

# Checking f1_score for BaggingClassifier

In [405]:
confusionmatrix=confusion_matrix(train_data_y,bag_model.predict(train_data_x))

In [406]:
precision=confusionmatrix[0][0]/(confusionmatrix[0][0]+confusionmatrix[0][1])
recall=confusionmatrix[0][0]/(confusionmatrix[0][0]+confusionmatrix[1][0])
f1_score_bagging=2*precision*recall/(precision+recall)
f1_score_bagging

0.9815622274822474

# Accuracy for BaggingClassifier

In [407]:
Accuracy_bagging=(confusionmatrix[0][0]+confusionmatrix[1][1])/(confusionmatrix[0][0]+confusionmatrix[1][1]+confusionmatrix[0][1]+confusionmatrix[1][0])
Accuracy_bagging

0.9662119742023857

# predicting target variable for testing dataset using Bagging classifier

In [409]:
bag_model_prediction=bag_model.predict(test_data_set[test_data_set.columns])
bag_model_prediction

array([0, 1, 1, 0])

# Using BoostingClassifier

* This method is used to **decrease the bias of the model** and **minimise the training errors**.

In [410]:
from sklearn.ensemble import AdaBoostClassifier
boost_model=AdaBoostClassifier(base_estimator=DecisionTreeClassifier(),n_estimators=10,learning_rate=0.2,random_state=0)
boost_model.fit(train_data_x,train_data_y)

AdaBoostClassifier(base_estimator=DecisionTreeClassifier(), learning_rate=0.2,
                   n_estimators=10, random_state=0)

# Checking f1_score for BoostingClassifier

In [411]:
confusionmatrix=confusion_matrix(train_data_y,boost_model.predict(train_data_x))

In [412]:
precision=confusionmatrix[0][0]/(confusionmatrix[0][0]+confusionmatrix[0][1])
recall=confusionmatrix[0][0]/(confusionmatrix[0][0]+confusionmatrix[1][0])
f1_score_boosting=2*precision*recall/(precision+recall)
f1_score_boosting

0.9958296474156451

# Accuracy for BoostingClassifier

In [413]:
Accuracy_boosting=(confusionmatrix[0][0]+confusionmatrix[1][1])/(confusionmatrix[0][0]+confusionmatrix[1][1]+confusionmatrix[0][1]+confusionmatrix[1][0])
Accuracy_boosting

0.9924661834370184

# predicting the target variable for testing dataset using Boosting classifier

In [414]:
boost_model_prediction=boost_model.predict(test_data_set[test_data_set.columns])
boost_model_prediction

array([0, 1, 1, 1])

# Applied models f1scores and accuracies with only 1774 independent features

In [415]:
model_dataset=pd.DataFrame({'Stacking classifier':[f1_score_stacking,Accuracy_stacking],'Bagging classifier':[f1_score_bagging,Accuracy_bagging],'Boosting Classifier':[f1_score_boosting,Accuracy_boosting]},index=['f1 score','Accuracy'])
model_dataset

Stacking classifier  Bagging classifier  Boosting Classifier
f1 score             0.995735            0.981562             0.995830
Accuracy             0.992295            0.966212             0.992466

# stacking model prediction for testing dataset values

In [416]:
stackmodel_prediction

array([1, 1, 1, 1])

# Bagging model prediction for testing dataset values

In [418]:
bag_model_prediction

array([0, 1, 1, 0])

# Boosting model prediction for testing dataset values

In [419]:
boost_model_prediction

array([0, 1, 1, 1])

# Conclusion
* With 1774 variables, we can observe that f1 and accuracy scores of both stacking and boosting classifier is >0.99 but we cannot say these models are highly preferable because they may be overfitted(may not predict testdata correctly)
* Bagging classifier with f1 score of 0.98 and accuracy score of 0.96 which can predict appropriately when compared to bagging and boosting model with only 1774 variables
* Bagging classifier predicted 2 values positively and 2 values negatively for the test dataset in approximate manner when compared to other models
* Bagging classifier is preferable to predict the credit risk

# BaggingClassifier is preferable as the f1_score=0.98,accuracy=0.96 with only 1774 independent features